src/porepy/numerics/nonlinear/convergence_check.py

```diff
@@ -0,0 +1,455 @@
+            for key, res_norm in residual_norm.items()
+        )
+
+        # Check convergence using relative norms.
+        converged_inc = reduced_nonlinear_increment_norm < tol.tol_increment
+        converged_res = reduced_residual_norm < tol.tol_residual
+        is_converged = converged_inc and converged_res
+
+        # Determine convergence status.
+        convergence_status = ConvergenceStatus.NOT_CONVERGED
+        if is_nan:
+            convergence_status = ConvergenceStatus.NAN
+        elif is_diverged:
+            convergence_status = ConvergenceStatus.DIVERGED
+        elif is_converged:
+            convergence_status = ConvergenceStatus.CONVERGED
+
+        # Collect information about the convergence check.
+        info = ConvergenceInfo(
+            nonlinear_increment_norm=reduced_nonlinear_increment_norm,
+            residual_norm=reduced_residual_norm,
+        )
+
+        return convergence_status, info
+
+
+### Concrete convergence criteria
+
+
+# class NanConvergenceCriterion(ConvergenceCriterion):
+#    """Convergence criterion that checks for NaN values."""
+#
+#    def _check(
+#        self,
+#        nonlinear_increment: dict[str, float],
+#        residual: dict[str, float],
+#        tol: ConvergenceTolerance,
+#    ) -> tuple[ConvergenceStatus, ConvergenceInfo]:
+#        """Check for NaN values in the nonlinear increment and residual.
+#
+#        Parameters:
+#            nonlinear_increment: The increment in the solution variables from the
+#                previous nonlinear iteration.
+#            residual: The current residual vector of the nonlinear system.
+#            tol: Not used, but required for complying with super class.
+#
+#        Returns:
+#            ConvergenceStatus: Convergence status of the non-linear iteration.
+#            ConverenceInfo: Information about the convergence check.
+#
+#        """
+#        has_nan_increment = any(
+#            np.isnan(value) for value in nonlinear_increment.values()
+#        )
+#        has_nan_residual = any(np.isnan(value) for value in residual.values())
+#        if has_nan_increment or has_nan_residual:
+#            return ConvergenceStatus.NAN, ConvergenceInfo(np.nan, np.nan)
+#        else:
+#            return ConvergenceStatus.CONVERGED, ConvergenceInfo(0.0, 0.0)
+
+
+class AbsoluteConvergenceCriterion(RelativeConvergenceCriterion):
+    """Absolute convergence criterion for nonlinear problems."""
+
+    def init_reference_value(self):
+        """Initialize the reference value manager for absolute convergence."""
+        self.reference_value = ReferenceValue(
+            condition=lambda x: False,
+            default_reference_value=0.0,
+        )
+
+
+class DynamicRelativeConvergenceCriterion(RelativeConvergenceCriterion):
+    """Relative convergence criterion for nonlinear problems.
+
+    Reference values are set based on the current state but are not allowed to
+    be zero or nan.
```

Comment:
Why not zero? This should be fine, given the 1+ref. I think I can sketch a scenario where setting to zero is preferable. Let me know if we should discuss.